service/src/main/java/uk/gov/hmcts/reform/fpl/controllers/MessageJudgeController.java

```diff
@@ -0,0 +1,77 @@
+import org.springframework.web.bind.annotation.RestController;
+import uk.gov.hmcts.reform.ccd.client.model.AboutToStartOrSubmitCallbackResponse;
+import uk.gov.hmcts.reform.ccd.client.model.CallbackRequest;
+import uk.gov.hmcts.reform.ccd.client.model.CaseDetails;
+import uk.gov.hmcts.reform.fpl.events.NewJudicialMessageEvent;
+import uk.gov.hmcts.reform.fpl.model.CaseData;
+import uk.gov.hmcts.reform.fpl.model.JudicialMessage;
+import uk.gov.hmcts.reform.fpl.model.common.Element;
+import uk.gov.hmcts.reform.fpl.service.MessageJudgeService;
+import uk.gov.hmcts.reform.fpl.utils.CaseDetailsMap;
+
+import java.util.List;
+
+import static uk.gov.hmcts.reform.fpl.utils.CaseDetailsHelper.removeTemporaryFields;
+import static uk.gov.hmcts.reform.fpl.utils.CaseDetailsMap.caseDetailsMap;
+
+@Api
+@RestController
+@RequestMapping("/callback/message-judge")
+@RequiredArgsConstructor(onConstructor = @__(@Autowired))
+public class MessageJudgeController extends CallbackController {
+    private final MessageJudgeService messageJudgeService;
+
+    @PostMapping("/about-to-start")
+    public AboutToStartOrSubmitCallbackResponse handleAboutToStart(@RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        CaseData caseData = getCaseData(caseDetails);
+        CaseDetailsMap caseDetailsMap = caseDetailsMap(caseDetails);
+
+        caseDetailsMap.putAll(messageJudgeService.initialiseCaseFields(caseData));
+        removeTemporaryFields(caseDetailsMap, "judicialMessageMetaData", "judicialMessageNote");
+
+        return respond(caseDetailsMap);
+    }
+
+    @PostMapping("/mid-event")
+    public AboutToStartOrSubmitCallbackResponse handleMidEvent(@RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        CaseData caseData = getCaseData(caseDetails);
+        CaseDetailsMap caseDetailsMap = caseDetailsMap(caseDetails);
+
+        caseData.getFirstHearing().ifPresent(hearingBooking -> caseDetailsMap.put("nextHearingLabel",
+            String.format("Next hearing in the case: %s", hearingBooking.toLabel())));
+
+        caseDetailsMap.putAll(messageJudgeService.buildRelatedC2DocumentFields(caseData));
+
+        return respond(caseDetailsMap);
+    }
+
+    @PostMapping("/about-to-submit")
+    public AboutToStartOrSubmitCallbackResponse handleAboutToSubmit(@RequestBody CallbackRequest callbackRequest) {
+        CaseDetails caseDetails = callbackRequest.getCaseDetails();
+        CaseData caseData = getCaseData(caseDetails);
+        CaseDetailsMap caseDetailsMap = caseDetailsMap(caseDetails);
+
+        List<Element<JudicialMessage>> updatedMessages = messageJudgeService.addNewJudicialMessage(caseData);
+        caseDetailsMap.put("judicialMessages", messageJudgeService.sortJudicialMessages(updatedMessages));
+
+        removeTemporaryFields(caseDetailsMap, "hasC2Applications", "isMessageRegardingC2", "c2DynamicList",
```

Comment:
these tranisent field could be added to the event data object that you made. maybe they could become fields there as well